File: pinochle/round.py
from itertools import cycle
from typing import List

from pinochle import Player, Deck, Hand


class Round:
    """ One round of play """

    def __init__(self, players: List[Player], dealer: Player):
        """ Starts a round """
        self.players: List[Player] = players
        self.dealer: Player = dealer

    def player_iterator(self):
        """ Creates an iterator for the players that follow the dealer """
        players = self.players
        dealer = self.dealer
        d = players.index(dealer)
        if d == -1:
            errmsg = f"Dealer {self.dealer} is not one of the players"
            raise ValueError(errmsg)
        d = (d + 1) % 4
        return cycle(self.players[d:] + self.players[0:d])

    def deal(self):
        """ Deals each player a hand """
        # Empty hands for each player to begin with
        players = self.players
        for player in players:
            player.hand = Hand()

        # Shuffle and deal
        deck = Deck()
        deck.shuffle()
        it = self.player_iterator()
        for card in deck.cards:
            player = next(it)
            player.hand.add(card)

File: round.go
package pinochle

// ---------------------------------------------------------------------
// Type Definitions
// ---------------------------------------------------------------------

// Round is one round of play
type Round struct {
	game      *Game
	dealer    *Player
	bidWinner *Player
	trump     Suit
}

// ---------------------------------------------------------------------
// Constructor
// ---------------------------------------------------------------------

// NewRound starts a new round
func NewRound(g *Game, dealer *Player) *Round {

	// Create a new round
	r := new(Round)
	r.game = g
	r.dealer = dealer

	// Create a new deck
	deck := NewDeck()

	// Dealer shuffles at least once
	deck.Shuffle()
	for dealer.WantToReshuffle() {
		deck.Shuffle()
	}

	// Dealer offers player on right chance to cut
	other := g.PlayerOnRight(dealer)
	if other.WantToCut() {
		deck.Cut()
	}

	// Deal the cards, three at a time
	for _, player := range g.Players {
		player.Hand = []Card{}
	}
	thisPlayer := dealer
	other = g.PlayerOnLeft(thisPlayer)
	for i, card := range deck.cards {
		other.Hand = append(other.Hand, card)
		if i % 4 == 3 {
			other = g.PlayerOnLeft(other)
		}
	}

	// Return a pointer to the round
	return r
}

// ---------------------------------------------------------------------
// Methods
// ---------------------------------------------------------------------
